README.md
# Recompile article

All figures have already been generated. They are in PDF format. To recompile
the paper:

    $ pdflatex switches_in_the_brain.tex 


# Regenerating everything again

This is not recommended. Please write to dilawars@ncbs.res.in if you need
something changed in figures.

If you are feeling bold, then you can do the following:

    $ make -B

To generate pdf your need to install `gnuplot`, latest texlive and `latexmk`. On
ubuntu machine:

    $ sudo apt install latexmk gnuplot


CHANGELOG
Sep 26, 2018

- Expanded conclusion.
- Added box to explain bistability.
- Added box to explain CAMKII.


July 05, 2018

- First submission.

bistable/strongBis.py
#########################################################################
## This program is part of 'MOOSE', the
## Messaging Object Oriented Simulation Environment.
##           Copyright (C) 2014 Upinder S. Bhalla. and NCBS
## It is made available under the terms of the
## GNU Lesser General Public License version 2.1
## See the file COPYING.LIB for the full notice.
# Saturday 22 September 2018 09:53:49 AM IST
#########################################################################

import moose
import matplotlib.pyplot as plt
import matplotlib.image as mpimg
import pylab
import numpy
import sys
import pandas as pd

def main():
        """
        This example illustrates loading, and running a kinetic model 
        for a bistable system, defined in kkit format. 
        Defaults to the deterministic gsl method, you can pick the 
        stochastic one by 

            ``python filename gssa``

        The model starts out equally poised between sides **b** and **c**. 
        Then there is a small molecular 'tap' to push it over to **b**.
        Then we apply a moderate push to show that it is now very stably in
        this state. it takes a strong push to take it over to **c**.
        Then it takes a strong push to take it back to **b**.
        This is a good model to use as the basis for running stochastically
        and examining how state stability is affected by changing volume.
        """

        solver = "gsl"  # Pick any of gsl, gssa, ee..
        #solver = "gssa"  # Pick any of gsl, gssa, ee..
        #moose.seed( 1234 ) # Needed if stochastic.
        mfile = './M1719.g'
        runtime = 300.0
        if ( len( sys.argv ) >= 2 ):
            solver = sys.argv[1]
        modelId = moose.loadModel( mfile, 'model', solver )
        # Increase volume so that the stochastic solver gssa 
        # gives an interesting output
        compt = moose.element( '/model/kinetics' )
        compt.volume = 0.2e-19 
        r = moose.element( '/model/kinetics/equil' )

        moose.reinit()
        moose.start( runtime ) 
        r.Kf *= 1.1 # small tap to break symmetry
        moose.start( runtime/10 ) 
        r.Kf = r.Kb
        moose.start( runtime ) 

        r.Kb *= 2.0 # Moderate push does not tip it back.
        moose.start( runtime/10 ) 
        r.Kb = r.Kf
        moose.start( runtime ) 

        r.Kb *= 5.0 # Strong push does tip it over
        moose.start( runtime/10 ) 
        r.Kb = r.Kf
        moose.start( runtime ) 
        r.Kf *= 5.0 # Strong push tips it back.
        moose.start( runtime/10 ) 
        r.Kf = r.Kb
        moose.start( runtime ) 

        # Display all plots.
        df = pd.DataFrame()
        img = mpimg.imread( 'strongBis.png' )
        fig = plt.figure( figsize=(12, 10 ) )
        png = fig.add_subplot( 211 )
        imgplot = plt.imshow( img )
        ax = fig.add_subplot( 212 )
        x = moose.wildcardFind( '/model/#graphs/conc#/#' )
        dt = moose.element( '/clock' ).tickDt[18]
        t = numpy.arange( 0, x[0].vector.size, 1 ) * dt
        df['time'] = t
        for i in range(3):
            df[x[i].name] = x[i].vector
        ax.plot( t, x[0].vector, 'r-', label=x[0].name )
        ax.plot( t, x[1].vector, 'g-', label=x[1].name )
        ax.plot( t, x[2].vector, 'b-', label=x[2].name )
        plt.ylabel( 'Conc (mM)' )
        plt.xlabel( 'Time (seconds)' )
        pylab.legend()
        pylab.savefig( 'res.png' )
        df.to_csv( '%s.csv' % solver, index = False )

# Run the 'main' if this script is executed standalone.
if __name__ == '__main__':
    main()


hopfield/hopfield_demo.py
import numpy as np
import random
import matplotlib.pyplot as plt
from neurodynex.hopfield_network import network, pattern_tools, plot_tools

add_extra = False

# the letters we want to store in the hopfield network
letter_list = ['N', 'C', 'B', 'S', 'X', 'Y' ]
extra = [ 'P', 'Q' ]
if add_extra == True:
    letter_list += extra

nRows, maxCols = 4, len(letter_list)
abc_dictionary = pattern_tools.load_alphabet()

def save_letter( a, filename ):
    with open( filename, 'w') as f:
        for (i, j), v in np.ndenumerate(a):
            f.write( '%d %d %d\n' % (i,j,v))
    print( '-- saved to %s' % filename )


def add_noise( letter, noise_level = 0.2 ):
    pat = abc_dictionary[letter]
    print( "[INFO ] Adding noise (level=%f) to %s" % (noise_level, letter))
    #  input( "Press any key to continue" )
    #  print( "  to %s" % letter )
    for (i, j), v in np.ndenumerate(pat):
        if random.random() < noise_level:
            pat[i, j] = random.choice( [-1, 1] )
    return pat

def main( plot = True ):
    # set a seed to reproduce the same noise in the next run
    # numpy.random.seed(123)
    
    # access the first element and get it's size (they are all of same size)
    pattern_shape = abc_dictionary['A'].shape

    # create an instance of the class HopfieldNetwork
    hopfield_net = network.HopfieldNetwork( 
            nr_neurons = pattern_shape[0]*pattern_shape[1]
            )
    
    pattern_list = [abc_dictionary[key] for key in letter_list ]

    if plot:
        plt.figure( figsize=(12, 6) )
        for i, k in enumerate(letter_list):
            a = abc_dictionary[k]
            ax = plt.subplot( nRows, maxCols, i+1)
            ax.imshow( a )
            ax.axis('off')
            save_letter( a, '%s.txt' % k )
    
    # store the patterns
    hopfield_net.store_patterns(pattern_list)
    print( "[INFO ] Saved patterns into hopfield network." )
    
    # create a noisy version of a pattern and use that to initialize the network
    l = random.choice( letter_list )

    noise_level = 0.5
    if add_extra:
        noise_level = 0.1
        print( '[INFO] More patterns than net can handle. Catastrophic'
            ' forgetting going to happen.'
        )
    noisy_init_state = add_noise( l, noise_level= noise_level)

    hopfield_net.set_state_from_pattern(noisy_init_state)
    
    # from this initial state, let the network dynamics evolve.
    states = hopfield_net.run_with_monitoring(nr_steps=4)

    # each network state is a vector. reshape it to the same shape used to create the patterns.
    states_as_patterns = pattern_tools.reshape_patterns(states, pattern_list[0].shape)

    if plot:
        for i, pat in enumerate( states_as_patterns ):
            overlap_list = pattern_tools.compute_overlap_list(pat, pattern_list)
            ax = plt.subplot( nRows, maxCols, maxCols+i+1 )
            save_letter( pat, '%s%s.txt' % (l, i) )
            ax.imshow( pat )
            ax.axis('off')
            ax1 = plt.subplot( nRows, maxCols, 2*maxCols+i+1)
            ax1.bar( range(len(overlap_list)), overlap_list )
            #  ax1.set_ylim( -1, 1 )
            ax1.set_xticks( range(len(overlap_list)) )
            ax1.set_xticklabels( letter_list )

        plt.tight_layout()
        outfile = 'figures/final_%s.png' % l
        plt.savefig( outfile )
        print( '|| Saved to %s' % outfile )
        plt.show()
        plt.close()

    return pattern_list, noisy_init_state, states_as_patterns

if __name__ == '__main__':
    main()


figures/integrator/main.py
"""main.py: 

"""
    
__author__           = "Dilawar Singh"
__copyright__        = "Copyright 2017-, Dilawar Singh"
__version__          = "1.0.0"
__maintainer__       = "Dilawar Singh"
__email__            = "dilawars@ncbs.res.in"
__status__           = "Development"

import sys
import os
import math
import matplotlib.pyplot as plt
import random
import numpy as np
import pandas as pd

def get_ca( t ):
    ca = 0.01
    if t > 15:
        return ca
    if t % 2 > 1:
        ca = 1 * (1 +  0.3 * random.gauss(0, 0.5))
    return ca

def leaky_integrate( xs, dt, tau ):
    val = [0.0]
    factor = 1e-3
    for x in xs:
        v = val[-1] + factor * x
        v -= v / tau    # decay
        val.append(v)
    return val[1:]

def main():
    dt = 0.01
    tvec = []
    ca = []
    df = pd.DataFrame()
    for t in np.arange( 0, 20.0, dt ):
        tvec.append(t )
        ca.append( get_ca( t ) )

    camkii0 = leaky_integrate( ca, dt, 1e9 )
    camkii1 = leaky_integrate( ca, dt, 1e3 )
    camkii2 = leaky_integrate( ca, dt, 2e2 )
    df['Time'] = tvec
    df['calcium'] = ca
    df['CaMKII0'] = camkii0
    df['CaMKII1'] = camkii1
    df['CaMKII2'] = camkii2

    ax = plt.subplot( 111 )
    ax.plot( tvec, ca , label = 'Ca')
    ax1 = ax.twinx()
    ax1.plot( tvec, camkii0, color = 'red' , label = 'CaMKII' )
    ax1.plot( tvec, camkii1, label = 'CaMKII' )
    ax1.plot( tvec, camkii2, label = 'CaMKII' )
    plt.savefig( '%s.png' % __file__ )

    df.to_csv( 'data.csv', index = False )

if __name__ == '__main__':
    main()


figures/hopfield/hopfield_demo_energy_landscape.py
"""hopfield_demo_energy_landscape.py: 

"""
    
__author__           = "Dilawar Singh"
__copyright__        = "Copyright 2017-, Dilawar Singh"
__version__          = "1.0.0"
__maintainer__       = "Dilawar Singh"
__email__            = "dilawars@ncbs.res.in"
__status__           = "Development"

import sys
import os
import matplotlib.pyplot as plt
import numpy as np
import hopfield_demo

def compute_energe( frame ):
    frame = (1+frame)/2.0
    return np.mean( frame )

def run(i, ax):
    print( "[INFO ] Running for %d" % i  )
    stored, clue, states = hopfield_demo.main( plot = False )
    traj = [ compute_energe(x) for x  in states ]
    ax.plot( traj[:3], 'o-' )

def main( ):
    plt.figure()
    ax = plt.subplot( 111 )
    for i in range(100):
        run(i, ax)

    plt.savefig( 'figures/energy_landscape.png' )

if __name__ == '__main__':
    main()


simulations/graded_bistable.py
"""milti_stable_bistable.py: 
"""
    
__author__           = "Dilawar Singh"
__copyright__        = "Copyright 2017-, Dilawar Singh"
__version__          = "1.0.0"
__maintainer__       = "Dilawar Singh"
__email__            = "dilawars@ncbs.res.in"
__status__           = "Development"

import sys
import os
import matplotlib.pyplot as plt
import numpy as np
import random
import pandas as pd

random.seed( 10 )

class Bistable( ):

    def __init__(self):
        self.pup = 0.01
        self.pdown = 0.01
        self.x = random.choice([0,1])
        self.xs = []

    def step( self ):
        self.xs.append( self.x )
        if self.x >= 0.8:
            if random.random() < self.pdown:
                self.x = random.uniform(0, 0.2)
            else:
                self.x = random.uniform(0.8, 1.0)
        elif self.x < 0.2:
            if random.random() < self.pup:
                self.x = random.uniform(0.8, 1.0)
            else:
                self.x = random.uniform(0, 0.2)
        else:
            pass

    def steps( self, N ):
        for i in range(N):
            self.step()

def main( ):
    a1 = Bistable( )
    a2 = Bistable( )
    a3 = Bistable( )
    a4 = Bistable( )
    a5 = Bistable( )
    N = 2000
    df = pd.DataFrame( )
    ax1 = plt.subplot(211)
    ax2 = plt.subplot(212)
    for i, a in enumerate([a1, a2, a3, a4, a5]):
        a.steps( int(N) )
        df[ 'switch%i' % i] = a.xs

    df.to_csv( 'bistable_graded.csv', index = False )
    ax1.plot( a1.xs )
    ax2.plot( np.sum( [a2.xs, a3.xs, a4.xs, a5.xs], axis = 0) )
    plt.savefig( '%s.png' % __file__ )

if __name__ == '__main__':
    main()
